File: src/notarize.js
"use strict";
// License notarization. Identify which SPDX license a file is, and attest the
// match — an offline, optionally issuer-signed "notary stamp". The attestation
// reuses the same Ed25519 receipt machinery as the deployment gate, so it
// verifies offline against the embedded public key.
const fs = require("fs");
const crypto = require("crypto");
const { getText, meta, list, version } = require("./catalogue");
const { signReceipt } = require("./sign"); // server-side stamp (optional)
const { verifyReceipt } = require("./verify");

function sha256(buf) {
  return crypto.createHash("sha256").update(buf).digest("hex");
}
// Whitespace-insensitive, case-insensitive normalization for "same license,
// different formatting" matching. (Filled-in placeholders still differ — that's
// reported as a non-exact match, by design.)
function normalize(s) {
  return String(s).replace(/\r\n/g, "\n").replace(/\s+/g, " ").trim().toLowerCase();
}

let _normIndex = null;
function normIndex() {
  if (_normIndex) return _normIndex;
  _normIndex = new Map();
  for (const l of list({ includeDeprecated: true })) {
    const t = getText(l.id);
    if (t != null) _normIndex.set(sha256(normalize(t)), l.id);
  }
  return _normIndex;
}

// notarize(filePath, { spdxId?, privateKeyPem? }) → attestation object
// (signed if a private key is supplied). match ∈ { exact, normalized, none }.
function notarize(filePath, { spdxId, privateKeyPem } = {}) {
  const content = fs.readFileSync(filePath);
  const fileSha = sha256(content);
  const fileNorm = sha256(normalize(content.toString("utf8")));

  let id = spdxId ? (meta(spdxId) ? meta(spdxId).licenseId : spdxId) : normIndex().get(fileNorm) || null;
  let match = "none", canonicalSha = null, name = null, osi = null;
  if (id) {
    const canon = getText(id);
    const m = meta(id);
    name = m ? m.name : null;
    osi = m ? !!m.isOsiApproved : null;
    if (canon != null) {
      canonicalSha = sha256(Buffer.from(canon, "utf8"));
      const canonNorm = sha256(normalize(canon));
      match = canonicalSha === fileSha ? "exact" : canonNorm === fileNorm ? "normalized" : "none";
    }
  }

  const att = {
    kind: "license-notarization",
    file: filePath,
    file_sha256: fileSha,
    file_normalized_sha256: fileNorm,
    spdx_id: id,
    license_name: name,
    osi_approved: osi,
    match,
    canonical_sha256: canonicalSha,
    spdx_list_version: version(),
    notary: "public-notary (Interchained LLC)",
    notarized_at: new Date().toISOString(),
  };
  return privateKeyPem ? signReceipt(att, privateKeyPem) : att;
}

// Verify a signed notarization stamp offline against the embedded public key.
function verifyAttestation(att) {
  return verifyReceipt(att);
}

module.exports = { notarize, verifyAttestation, sha256, normalize };

File: rust/src/notarize.rs
//! License notarization — identify a file's SPDX license and verify signed
//! attestations offline. Signing the "notary stamp" is server-side (JS/Python);
//! this crate identifies and verifies.
use crate::verify::verify_receipt;
use serde_json::Value;
use sha2::{Digest, Sha256};

pub fn sha256_hex(bytes: &[u8]) -> String {
    let mut h = Sha256::new();
    h.update(bytes);
    hex::encode(h.finalize())
}

/// Whitespace-insensitive, lowercased normalization (matches the JS port).
pub fn normalize(s: &str) -> String {
    s.replace("\r\n", "\n")
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ")
        .to_lowercase()
}

/// Verify a server-signed notarization attestation offline against the issuer key.
pub fn verify_attestation(att: &Value) -> bool {
    verify_receipt(att)
}

/// Identify which SPDX license `path` is and report the match. The returned
/// attestation is UNSIGNED — the signed stamp is issued by the license server.
/// `match` ∈ "exact" | "normalized" | "none".
pub fn notarize(path: &str, spdx_id: Option<&str>) -> std::io::Result<Value> {
    let content = std::fs::read(path)?;
    let file_sha = sha256_hex(&content);
    let text = String::from_utf8_lossy(&content);
    let file_norm = sha256_hex(normalize(&text).as_bytes());

    let id: Option<String> = match spdx_id {
        Some(s) => Some(s.to_string()),
        None => crate::catalogue::identify(&file_norm),
    };
    let mut matched = "none".to_string();
    let mut canonical_sha = Value::Null;
    let mut name = Value::Null;
    if let Some(ref id) = id {
        if let Some(canon) = crate::catalogue::get_text(id) {
            let csha = sha256_hex(canon.as_bytes());
            let cnorm = sha256_hex(normalize(&canon).as_bytes());
            matched = if csha == file_sha {
                "exact".to_string()
            } else if cnorm == file_norm {
                "normalized".to_string()
            } else {
                "none".to_string()
            };
            canonical_sha = Value::String(csha);
        }
        if let Some(nm) = crate::catalogue::name(id) {
            name = Value::String(nm);
        }
    }
    Ok(serde_json::json!({
        "kind": "license-notarization",
        "file": path,
        "file_sha256": file_sha,
        "file_normalized_sha256": file_norm,
        "spdx_id": id,
        "license_name": name,
        "match": matched,
        "canonical_sha256": canonical_sha,
        "notary": "public-notary (Interchained LLC)"
    }))
}
